### PR TITLE
Feature toml configs

### PR DESCRIPTION
**Pull request by kevin-harrison**
Please make sure these boxes are checked, before submitting a new PR.

- [x] You ran the local CI checker with `./check.sh` with no errors
- [x] You updated the OmniPaxos book (if applicable)

## Issues
Switches from Hocon configuration files to TOML.

## Other Changes
Removes unused `OmniPaxosConfig` field `logger_path`.

### CODE REVIEW

File: omnipaxos_core/Cargo.toml [hunk 1]
```diff
@@ -26,10 +26,9 @@ rand = "0.8.4"
...
-
-default = ["continued_leader_reconfiguration", "batch_accept", "hocon_config"]
+logging  = ["slog", "slog-term", "slog-async"]
+toml-config = ["serde", "toml"]
+default = ["continued_leader_reconfiguration", "batch_accept"]
 
 #[profile.release]
-#lto = true
\ No newline at end of file
+#lto = true
```
Comment (haraldng) [inline omnipaxos_core/Cargo.toml]: I know you didn't comment these, but they should actually be uncommented.

**Comment by Max-Meldrum** [comment]
The following is also possible. However, if ``toml-config`` is the preferred name, then that works too 😄 

```toml

[features]
toml = ["dep:toml", "serde"]

[dependencies]
serde = { version = "1.0", optional = true, features = ["derive"] }
toml = { version = "0.7.3", optional = true }
```

File: omnipaxos_core/Cargo.toml [hunk 2]
```diff
@@ -26,10 +26,9 @@ rand = "0.8.4"
 [features]
 batch_accept = []
 continued_leader_reconfiguration = []
-logging  = [ "slog", "slog-term", "slog-async"]
-hocon_config = [ "hocon" ]
-
-default = ["continued_leader_reconfiguration", "batch_accept", "hocon_config"]
+logging  = ["slog", "slog-term", "slog-async"]
+toml-config = ["serde", "toml"]
```
Comment (haraldng) [inline omnipaxos_core/Cargo.toml:30]: Let's be consistent with the other features in the library and use underscore.

```suggestion
toml_config = ["serde", "toml"]
```

File: omnipaxos_core/src/omni_paxos.rs [hunk 3]
```diff
@@ -22,49 +24,27 @@ use std::ops::RangeBounds;
...
-        if let Some(b) = h[BUFFER_SIZE].as_i64() {
-            config.buffer_size = b as usize;
-        }
-        if let Some(p) = h[PRIORITY].as_i64().map(|p| p as u64) {
-            config.leader_priority = p;
-        }
-        config
+    /// Creates a new `OmniPaxosConfig` from a `toml` file.
+    #[cfg(feature = "toml-config")]
+    pub fn with_toml(file_path: &str) -> Result<Self, Box<dyn Error>> {
```
Comment (haraldng) [inline omnipaxos_core/src/omni_paxos.rs:44]: Add custom error message.

File: omnipaxos_core/tests/config/node2.toml
Comment (haraldng) [inline omnipaxos_core/tests/config/node2.toml:1]: What is this used for?

File: omnipaxos_core/tests/config/node1.toml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+configuration_id = 1
+pid = 1
+peers = [2, 3]
+buffer_size = 100000
+logger_file_path = "/sequencepaxos/logs"
+leader_priority = 2
+skip_prepare_use_leader = { n = 5, priority = 2, pid = 2 }
+initial_leader = { n = 1, priority = 1, pid = 1 }
+logger_path = "/sequencepaxos/logs"
```
Comment (haraldng) [inline omnipaxos_core/tests/config/node1.toml:9]:
```suggestion
logger_file_path = "/omnipaxos/logs"
leader_priority = 2
skip_prepare_use_leader = { n = 5, priority = 2, pid = 2 }
initial_leader = { n = 1, priority = 1, pid = 1 }
logger_path = "/omnipaxos/logs"
```

File: omnipaxos_core/tests/utils.rs [hunk 5]
```diff
@@ -23,56 +24,53 @@ const COMMITLOG: &str = "/commitlog/";
...
+/// the configuration file `/tests/config/test.toml` using toml
+#[derive(Deserialize)]
+#[serde(default)]
 pub struct TestConfig {
-    pub wait_timeout: Duration,
     pub num_threads: usize,
     pub num_nodes: usize,
-    pub election_timeout: Duration,
+    pub wait_timeout_sec: u64,
+    pub election_timeout_ms: u64,
```
Comment (haraldng) [inline omnipaxos_core/tests/utils.rs:38]: let's have everything in ms for simplicity

**Review by haraldng — CHANGES_REQUESTED** [review]
Looks good. Please rename the feature and fix other minor comments.

File: omnipaxos_core/tests/config/node2.toml
Comment (kevin-harrison) [reply omnipaxos_core/tests/config/node2.toml:1]: the `config_some_fields_test` to make sure unspecified fields are deserialized to default values.

Comment (haraldng) [reply omnipaxos_core/tests/config/node2.toml:1]: Ok, comment it in file so others also know that.

**Comment by haraldng** [comment]
Also, add description of `toml_config` here: https://github.com/haraldng/omnipaxos/blob/master/omnipaxos_core/src/lib.rs#L6

(some of them are outdated and serde is missing. Fix them if you want, otherwise I can do it later)

File: omnipaxos_core/tests/config/node2.toml [hunk 6]
```diff
@@ -1 +1,2 @@
+# This file is used for tests in /tests/config_test.rs
```
Comment (haraldng) [inline omnipaxos_core/tests/config/node2.toml:1]:
```suggestion
# This file is used for testing that unspecified fields are deserialized to default values in /tests/config_test.rs 
```

**Review by haraldng — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -41,7 +41,7 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: test
-          args: --verbose --features default,logging,hocon_config
+          args: --verbose --features default,logging,toml_config
```
[the remark this answers is not in the document]
Comment (kevin-harrison) [reply .github/workflows/ci.yml:44]: Yeah we can remove it, but leaving it out reduces test coverage.

[the remark this answers is not in the document]
Comment (kevin-harrison) [reply .github/workflows/ci.yml:44]: Should we remove all non-default features or just toml_config?

[the remark this answers is not in the document]
Comment (haraldng) [reply .github/workflows/ci.yml:44]: Ah ok, as long as `cargo test` still works we can keep these here.

**Pull request merged**